### PR TITLE
Improve default `OS`'s CPU count getter

### PR DESCRIPTION
**Pull request by RandomShaper**
Probably this allows to get rid of the OS-specific implementations in some/all platforms, but that's another story...

### CODE REVIEW

**Comment by fire** [comment]
Is it possible to the logical core count?

**Comment by akien-mga** [comment]
> Probably this allows to get rid of the OS-specific implementations in some/all platforms, but that's another story...

I think this would be worth doing as part of this PR because as it stands, it doesn't impact anything AFAIK.

`OS_Unix` uses:
```cpp
int OS_Unix::get_processor_count() const {
        return sysconf(_SC_NPROCESSORS_CONF);
}
```
and `OS_Windows` uses:
```
int OS_Windows::get_processor_count() const {
        SYSTEM_INFO sysinfo;
        if (is_wow64()) {
                GetNativeSystemInfo(&sysinfo);
        } else {
                GetSystemInfo(&sysinfo);
        }

        return sysinfo.dwNumberOfProcessors;
}
```
which would override this.

Some platforms supersede it further but AFAIK they all extend either `OS_Unix` or `OS_Windows` so this change doesn't change anything.

I get that the purpose is likely for a third-party platform port but I think it's better if we can improve the overall code quality here (otherwise the third-party platform port can also just redefine this in its `OS_<name>`).

**Comment by RandomShaper** [comment]
As I see it, we have two options that make more sense than the current:
a. Make `get_processor_count()` abstract and just ket every platform implement it.
b. Keep the default implementation, but make it more faithfully, as this PR makes it. I'd say the reason for just returning `1` was that there didn't use to be a standard way to get the core count.

**Comment by akien-mga** [comment]
The point I'm trying to make is that the new default implementation might be *better* than the one in `OS_Unix` and `OS_Windows`, which are the only ones that will be used outside of thirdparty ports that wouldn't implement it themselves.

So while this is a fine improvement as is, I think we'd miss an opportunity for checking what `std::thread::hardware_concurrency()` does and whether it's better than our own platform-specific implementations. If it is, we should just remove the `OS_Unix` and `OS_Windows` overrides and use `std::thread::hardware_concurrency()` for all platforms consistently.

**Comment by akien-mga** [comment]
For the record this seems to be GCC's implementation:

https://github.com/gcc-mirror/gcc/blob/16e2427f50c208dfe07d07f18009969502c25dc8/libstdc%2B%2B-v3/src/c%2B%2B11/thread.cc#L44-L70

I'm not familiar with all this but it does include something similar to our `sysconf(_SC_NPROCESSORS_CONF)` as a fallback, though it queries `_SC_NPROCESSORS_ONLN` (i.e. the online processors, which may differ from the configured amount https://www.gnu.org/software/libc/manual/html_node/Processor-Resources.html).
I'd say it does look more accurate.

LLVM's implementation:

https://github.com/llvm/llvm-project/blob/a3a9b0743e317c8ada9f53efb9c3b5928aec77b8/libcxx/src/thread.cpp#L73-L99

Also uses `_SC_NPROCESSORS_ONLN` for Unix, and `GetSystemInfo().dwNumberOfProcessors` on Win32. It doesn't have the `is_wow64` check we use though so there's possibly a difference in behavior here. https://learn.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getsysteminfo

Also not sure how GCC on Windows would behave here. I assume it would use the above GCC implementation and thus rely on winpthread to get the concurrency count, instead of calling the `GetSystemInfo()` API.

So from this quick check it sounds like we could remove the `OS_Unix` override and keep the `OS_Windows` one?

**Comment by RandomShaper** [comment]
Microsoft's C++ library implementation is this:
https://github.com/microsoft/STL/blob/2f03bdf361f7f153b4216c60a0d9491c0be13a73/stl/inc/thread#L150-L152
https://github.com/microsoft/STL/blob/89b7b1254a2f56dcb0f421df43d772874e6de6c6/stl/src/cthread.cpp#L105-L109

So, the `is_wow64()` check doesn't seem to be necessary.

Regarding `_SC_NPROCESSORS_ONLN`, a quick search revealed that the concept of online processors applies to systems where CPUs are hot pluggable! Not a conventional scenario.

About libc++ on Windows, under the believe that they are using pthread's implementation, I've checked the docs for `pthread_num_processors_np()`, which says that the number of processors returned may be lower than the actual, depending on the process core affinity mask. I'd say we can go with that.

In conclusion, I agree that it's a safe bet to keep the `std::hardware_concurrency` based implementation and remove the others. I'm updating the PR just now.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by timothyqiu** [comment]
Cherry-picked for 3.6
